File: binary_tree_count_number_of_nodes/binary_tree_count_number_of_nodes.java
/*
 * This program finds ithe number of nodes in a binary tree.
 */

import java.io.*;
import java.util.*;

/*
 * The class definition of a binary tree node
 */
class treenode
{
    int data;
    treenode left;
    treenode right;

    public treenode(int val)
    {
        data = val;
        left = null;
        right = null;
    }
}

public class binary_tree_count_number_of_nodes
{
    /*
     * This function recursively computes the number of nodes in a binary tree.
     */
    public static int find_number_of_nodes_binary_tree (treenode root)
    {
        /*
         * If the root node is null, then return zero.
         */
        if (root == null) {
            return(0);
        }

        /*
         * Otherwise return the sum of the following:-
         * 1. '1' for the node itself.
         * 2. number of nodes in the left sub-tree
         * 3. number of nodes in the right sub-tree
         */
        return(1 + find_number_of_nodes_binary_tree(root.left)
               + find_number_of_nodes_binary_tree(root.right));
    }

    public static void main (String[] args)
    {
        /*
         * Test 1: In case root is null, the number of nodes in the
         * tree should be zero.
         */
        treenode root1 = null;
        assert 0 == find_number_of_nodes_binary_tree(root1);

        /*
         * Test2: Test a bsic case of root with two child nodes.
         */
        treenode root2 = null;
        root2 = new treenode(10);
        root2.left = new treenode(20);
        root2.right = new treenode(30);
        assert 3 == find_number_of_nodes_binary_tree(root2);

        /*
         * Test 3: Test a case when the binary tree is a complete binary
         *         tree.
         */
        treenode root3 = null;
        root3 = new treenode(10);
        root3.left = new treenode(20);
        root3.right = new treenode(30);
        root3.left.right = new treenode(40);
        root3.right.right = new treenode(50);
        assert 5 == find_number_of_nodes_binary_tree(root3);

        /*
         * Test 4: Test a case when the tree is not a complete binary tree.
         */
        treenode root4 = null;
        root4 = new treenode(10);
        root4.left = new treenode(20);
        root4.right = new treenode(30);
        root4.left.left = new treenode(40);
        root4.left.right = new treenode(50);
        root4.right.left = new treenode(60);
        root4.right.left.right = new treenode(70);
        root4.right.left.left = new treenode(80);
        assert 8 == find_number_of_nodes_binary_tree(root4);
    }
}

File: binary_tree_count_number_of_nodes/binary_tree_count_number_of_nodes.c
/*
 * This program finds the number of nodes in a binary tree.
 */
#include<stdio.h>
#include<stdlib.h>
#include<assert.h>

/*
 * Structure definition for tree node
 */
struct treenode
{
    int data;
    struct treenode* left;
    struct treenode* right;
};

/*
 * This function allocates a tree node and sets the data field
 * in the node to be equal to 'val'.
 */
struct treenode* makenode (int val)
{
    struct treenode* new_node =
            (struct treenode*)malloc(sizeof(struct treenode));

    if (!new_node)
        return(NULL);

    new_node->data = val;
    new_node->left = NULL;
    new_node->right = NULL;

    return(new_node);
}

/*
 * This function recursively computes the number of nodes in a binary tree.
 * The time complexity of this function is O(n) since we are visiting each
 * node once.
 */
int find_number_of_nodes_binary_tree (struct treenode* root)
{
    /*
     * If the root node is NULL, then return zero.
     */
    if (!root) {
        return(0);
    }

    /*
     * Otherwise return the sum of the following:-
     * 1. '1' for the node itself.
     * 2. number of nodes in the left sub-tree
     * 3. number of nodes in the right sub-tree
     */
    return(1 + find_number_of_nodes_binary_tree(root->left)
           + find_number_of_nodes_binary_tree(root->right));
}

int main()
{
    /*
     * Test 1: In case root is NULL, the number of nodes in the
     * tree should be zero.
     */
    struct treenode* root1 = NULL;
    assert(0 == find_number_of_nodes_binary_tree(root1));

    /*
     * Test2: Test a bsic case of root with two child nodes.
     */
    struct treenode* root2 = NULL;
    root2 = makenode(10);
    root2->left = makenode(20);
    root2->right = makenode(30);
    assert(3 == find_number_of_nodes_binary_tree(root2));

    /*
     * Test 3: Test a case when the binary tree is a complete binary
     *         tree.
     */
    struct treenode* root3 = NULL;
    root3 = makenode(10);
    root3->left = makenode(20);
    root3->right = makenode(30);
    root3->left->right = makenode(40);
    root3->right->right = makenode(50);
    assert(5 == find_number_of_nodes_binary_tree(root3));

    /*
     * Test 4: Test a case when the tree is not a complete binary tree.
     */
    struct treenode* root4 = NULL;
    root4 = makenode(10);
    root4->left = makenode(20);
    root4->right = makenode(30);
    root4->left->left = makenode(40);
    root4->left->right = makenode(50);
    root4->right->left = makenode(60);
    root4->right->left->right = makenode(70);
    root4->right->left->left = makenode(80);
    assert(8 == find_number_of_nodes_binary_tree(root4));

    return(0);
}
